File: full-Stack-BP/full-Stack-BP/src/main/java/com/example/full_Stack_BP/services/TransactionService.java
package com.example.full_Stack_BP.services;

import com.example.full_Stack_BP.dto.request.TransactionRequestDTO;
import com.example.full_Stack_BP.dto.response.TransactionResponseDTO;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

public interface TransactionService {
    TransactionResponseDTO createTransaction(TransactionRequestDTO transactionRequest);
    TransactionResponseDTO deposit(TransactionRequestDTO transactionRequest);
    TransactionResponseDTO withdraw(TransactionRequestDTO transactionRequest);
    TransactionResponseDTO transfer(TransactionRequestDTO transactionRequest);  // Nuevo método
    TransactionResponseDTO getTransactionById(Long id);
    List<TransactionResponseDTO> getTransactionsByAccount(String accountNumber);
    List<TransactionResponseDTO> getTransactionsByClientAndDateRange(
            Long clientId, LocalDateTime startDate, LocalDateTime endDate);
    BigDecimal getDailyWithdrawalTotal(Long clientId);
}

File: full-stack-bp-frontend/src/app/services/transaction.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface Transaction {
  id?: number;
  date: string;
  transactionType: string;
  amount: number;
  balance: number;
  description: string;
  account: {
    id: number;
    accountNumber: string;
    accountType: string;
    client: {
      id: number;
      name: string;
      identification: string;
    } | null;
  };
}

export interface TransactionRequest {
  accountNumber: string;
  transactionType: string;
  amount: number;
  description: string;
  destinationAccountNumber?: string;
}

@Injectable({
  providedIn: 'root'
})
export class TransactionService {
  private apiUrl = `${environment.apiUrl}/transactions`;

  constructor(private http: HttpClient) { }

  createTransaction(request: TransactionRequest): Observable<Transaction> {
    return this.http.post<Transaction>(this.apiUrl, request);
  }

  deposit(request: TransactionRequest): Observable<Transaction> {
    return this.http.post<Transaction>(`${this.apiUrl}/deposit`, request);
  }

  withdraw(request: TransactionRequest): Observable<Transaction> {
    return this.http.post<Transaction>(`${this.apiUrl}/withdraw`, request);
  }

  transfer(request: TransactionRequest): Observable<Transaction> {
    return this.http.post<Transaction>(`${this.apiUrl}/transfer`, request);
  }

  getAllTransactionsByAccount(accountNumber: string): Observable<Transaction[]> {
    return this.http.get<Transaction[]>(`${this.apiUrl}/account/${accountNumber}`);
  }

  getTransactionById(id: number): Observable<Transaction> {
    return this.http.get<Transaction>(`${this.apiUrl}/${id}`);
  }

  getTransactionsByClientAndDateRange(
    clientId: number,
    startDate: string,
    endDate: string
  ): Observable<Transaction[]> {
    return this.http.get<Transaction[]>(
      `${this.apiUrl}/client/${clientId}?startDate=${startDate}&endDate=${endDate}`
    );
  }
}